File: blormlib/src/main/java/br/com/bloder/blormlib/validation/validations/Number.java
package br.com.bloder.blormlib.validation.validations;

import android.text.TextUtils;
import android.widget.EditText;

import br.com.bloder.blormlib.validation.Validate;
import br.com.bloder.blormlib.validation.Validation;

/**
 * Created by bloder on 31/07/16.
 */
public class Number extends Validation {

  private int maxSize = -1;
  private int minSize = -1;
  private boolean isOnlyPositive = false;
  private boolean isOnlyNegative = false;

  public Number withMaxSize(int maxSize) {
    this.maxSize = maxSize;
    return this;
  }

  public Number withMinimumSize(int minSize) {
    this.minSize = minSize;
    return this;
  }

  public Number andPositive() {
    isOnlyPositive = true;
    isOnlyNegative = false;
    return this;
  }

  public Number andNegative() {
    isOnlyNegative = true;
    isOnlyPositive = false;
    return this;
  }

  EditText editText;

  @Override
  public Validate validate() {
    return new Validate() {
      @Override
      public boolean validate() {
        editText = (EditText) field;
        return checkNumber(editText);
      }

      @Override
      public void onError() {
        editText.setError(errorMessage != null ? errorMessage : "This field only accept numbers!");
      }

      @Override
      public void onSuccess() {

      }
    };
  }

  private boolean checkNumber(EditText editText) {
    if (editText.getText().toString().isEmpty()) {
      return false;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && isOnlyPositive && !isOnlyNegative && maxSize == -1 && minSize == -1) {
      return Integer.parseInt(editText.getText().toString()) > 0;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString().substring(1)) && isOnlyNegative && !isOnlyPositive && maxSize == -1 && minSize == -1) {
      return String.valueOf(editText.getText().charAt(0)).equals("-") && editText.getText().toString().length() > 1 && TextUtils.isDigitsOnly(editText.getText().toString().substring(1));
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && (maxSize > 0 && minSize > 0) && (!isOnlyNegative && !isOnlyPositive)) {
      return editText.getText().toString().length() >= minSize && editText.getText().toString().length() <= maxSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && (maxSize == -1 && minSize >= 0) && (!isOnlyNegative && !isOnlyPositive)) {
      return editText.getText().toString().length() >= minSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && (maxSize >= 0 && minSize == -1) && (!isOnlyNegative && !isOnlyPositive)) {
      return editText.getText().toString().length() <= maxSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && isOnlyPositive && !isOnlyNegative && maxSize >= 0 && minSize >= 0) {
      return Integer.parseInt(editText.getText().toString()) > 0 && editText.getText().toString().length() >= minSize && editText.getText().toString().length() <= maxSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString().substring(1)) && isOnlyNegative && !isOnlyPositive && maxSize >= 0 && minSize >= 0) {
      return String.valueOf(editText.getText().charAt(0)).equals("-")
              && editText.getText().toString().length() > 1
              && TextUtils.isDigitsOnly(editText.getText().toString().substring(1))
              && editText.getText().toString().substring(1).length() >= minSize
              && editText.getText().toString().substring(1).length() <= maxSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && isOnlyPositive && !isOnlyNegative && maxSize == -1 && minSize >= 0) {
      return Integer.parseInt(editText.getText().toString()) > 0 && editText.getText().toString().length() >= minSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && isOnlyPositive && !isOnlyNegative && maxSize >= 0 && minSize == -1) {
      return Integer.parseInt(editText.getText().toString()) > 0 && editText.getText().toString().length() <= maxSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString().substring(1)) && isOnlyNegative && !isOnlyPositive && maxSize == -1 && minSize >= 0) {
      return String.valueOf(editText.getText().charAt(0)).equals("-")
              && editText.getText().toString().length() > 1
              && TextUtils.isDigitsOnly(editText.getText().toString().substring(1))
              && editText.getText().toString().substring(1).length() >= minSize;
    } else if (TextUtils.isDigitsOnly(editText.getText().toString()) && isOnlyNegative && !isOnlyPositive && maxSize >= 0 && minSize == -1) {
      return String.valueOf(editText.getText().charAt(0)).equals("-")
              && editText.getText().toString().length() > 1
              && TextUtils.isDigitsOnly(editText.getText().toString().substring(1))
              && editText.getText().toString().substring(1).length() <= maxSize;
    }
    return TextUtils.isDigitsOnly(editText.getText().toString());
  }
}

File: blormlib/src/main/java/br/com/bloder/blormlib/validation/validations/Number.kt
package br.com.bloder.blormlib.validation.validations

import android.text.TextUtils
import android.widget.EditText
import br.com.bloder.blormlib.validation.Validate
import br.com.bloder.blormlib.validation.Validation

/**
 * Created by bloder on 03/09/16.
 */
class Number : Validation(), Validate {

    var editText: EditText? = null
    private var maxSize = -1
    private var minSize = -1
    private var isOnlyPositive = false
    private var isOnlyNegative = false

    override fun validate(): Validate = this
    override fun validateIt(): Boolean {
        editText = field as EditText
       return checkNumber(editText!!)
    }
    override fun onError() { editText?.error = errorMessage ?: "This field only accept numbers!" }
    override fun onSuccess() { editText?.error = null }

    fun withMaxSize(maxSize: Int): Number {
        this.maxSize = maxSize
        return this
    }

    fun withMinimumSize(minSize: Int): Number {
        this.minSize = minSize
        return this
    }

    fun andPositive(): Number {
        isOnlyPositive = true
        isOnlyNegative = false
        return this
    }

    fun andNegative(): Number {
        isOnlyNegative = true
        isOnlyPositive = false
        return this
    }

    fun checkNumber(editText: EditText): Boolean {
        if (editText?.text.toString().isEmpty()) {
            return false
        } else if (TextUtils.isDigitsOnly(editText.text.toString()) && isOnlyPositive && !isOnlyNegative && maxSize == -1 && minSize == -1) {
            return Integer.parseInt(editText.text.toString()) > 0
        } else if (TextUtils.isDigitsOnly(editText.text.toString().substring(1)) && isOnlyNegative && !isOnlyPositive && maxSize == -1 && minSize == -1) {
            return editText?.text[0].toString() == "-" && editText.text.toString().length > 1 && TextUtils.isDigitsOnly(editText.text.toString().substring(1))
        } else if (TextUtils.isDigitsOnly(editText.text.toString()) && maxSize > 0 && minSize > 0 && !isOnlyNegative && !isOnlyPositive) {
            return editText?.text.toString().length >= minSize && editText.text.toString().length <= maxSize
        } else if (TextUtils.isDigitsOnly(editText.text.toString()) && maxSize == -1 && minSize >= 0 && !isOnlyNegative && !isOnlyPositive) {
            return editText?.text.toString().length >= minSize
        } else if (TextUtils.isDigitsOnly(editText.text.toString()) && maxSize >= 0 && minSize == -1 && !isOnlyNegative && !isOnlyPositive) {
            return editText?.text.toString().length <= maxSize
        } else if (TextUtils.isDigitsOnly(editText.text.toString()) && isOnlyPositive && !isOnlyNegative && maxSize >= 0 && minSize >= 0) {
            return Integer.parseInt(editText.text.toString()) > 0 && editText.text.toString().length >= minSize && editText.text.toString().length <= maxSize
        } else if (TextUtils.isDigitsOnly(editText.text.toString().substring(1)) && isOnlyNegative && !isOnlyPositive && maxSize >= 0 && minSize >= 0) {
            return editText?.text[0].toString() == "-"
                    && editText?.text.toString().length > 1
                    && TextUtils.isDigitsOnly(editText?.text.toString().substring(1))
                    && editText?.text.toString().substring(1).length >= minSize
                    && editText?.text.toString().substring(1).length <= maxSize
        } else if (TextUtils.isDigitsOnly(editText?.text.toString()) && isOnlyPositive && !isOnlyNegative && maxSize == -1 && minSize >= 0) {
            return Integer.parseInt(editText?.text.toString()) > 0 && editText.text.toString().length >= minSize
        } else if (TextUtils.isDigitsOnly(editText?.text.toString()) && isOnlyPositive && !isOnlyNegative && maxSize >= 0 && minSize == -1) {
            return Integer.parseInt(editText?.text.toString()) > 0 && editText.text.toString().length <= maxSize
        } else if (TextUtils.isDigitsOnly(editText?.text.toString().substring(1)) && isOnlyNegative && !isOnlyPositive && maxSize == -1 && minSize >= 0) {
            return editText?.text[0].toString() == "-"
                    && editText?.text.toString().length > 1
                    && TextUtils.isDigitsOnly(editText?.text.toString().substring(1))
                    && editText?.text.toString().substring(1).length >= minSize
        } else if (TextUtils.isDigitsOnly(editText?.text.toString()) && isOnlyNegative && !isOnlyPositive && maxSize >= 0 && minSize == -1) {
            return editText?.text[0].toString() == "-"
                    && editText?.text.toString().length > 1
                    && TextUtils.isDigitsOnly(editText?.text.toString().substring(1))
                    && editText?.text.toString().substring(1).length <= maxSize
        }
        return TextUtils.isDigitsOnly(editText?.text.toString())
    }
}
